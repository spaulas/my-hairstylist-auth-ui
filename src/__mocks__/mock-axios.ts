export const mockPost = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

export const mockGet = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

export const mockPut = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

export const mockDelete = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

export const mockPatch = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

const mockAxios = jest.fn().mockReturnValue({
  post: mockPost,
  get: mockGet,
  put: mockPut,
  patch: mockPatch,
  delete: mockDelete,
});

jest.mock("../config/HttpClient/index.ts", () => ({
  __esModule: true,
  default: mockAxios,
}));
