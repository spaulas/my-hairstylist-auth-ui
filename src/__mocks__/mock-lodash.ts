export const mockDebounce = jest.fn((fn) => console.log('MOCK DEBOUTNCE'));

jest.mock("lodash.debounce", () => ({
  ...jest.requireActual("lodash.debounce"),
  debounce: mockDebounce,
}));
