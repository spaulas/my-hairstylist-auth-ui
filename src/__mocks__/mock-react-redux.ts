import { ExplicitAny } from "@type/global";

export const mockSelector = jest.fn();

export const mockDispatch = jest.fn().mockImplementation((action: ExplicitAny) => null);

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useSelector: mockSelector,
    useDispatch: () => mockDispatch
}));
