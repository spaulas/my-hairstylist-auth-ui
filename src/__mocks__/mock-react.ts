export const mockUseState = jest.fn();
export const mockUseCallBack = jest.fn((fn) => fn);

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: mockUseState,
  useCallback: mockUseCallBack,
}));
