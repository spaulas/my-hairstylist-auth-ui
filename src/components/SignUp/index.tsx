import React, { useState } from "react";
import Form from "@components/generics/Form";
import EmailField from "@components/generics/Form/Field/Email";
import PasswordField from "@components/generics/Form/Field/Password";
import NameField from "@components/generics/Form/Field/Name";
import useAuth from "@hooks/useAuth/index.hooks";

const SignUp = () => {
  const { signUp } = useAuth();
  const [fields, setFields] = useState({ name: "", password: "", email: "" });

  const handleSignUp = () => {
    signUp(fields);
  };

  return (
    <Form
      handleSubmit={handleSignUp}
      buttonTitle="Sign up"
      link={{
        message: "Already have an account?",
        title: "Sign in",
        url: "/auth/sign-in",
      }}
    >
      {(isFocused: any, setIsFocused: any) => (
        <>
          <NameField
            name={fields.name}
            setName={(value: string) => setFields({ ...fields, name: value })}
            isFocused={isFocused}
            setIsFocused={(value: any) =>
              setIsFocused({ ...isFocused, name: value })
            }
          />
          <EmailField
            email={fields.email}
            setEmail={(value: string) => setFields({ ...fields, email: value })}
            isFocused={isFocused}
            setIsFocused={(value: any) =>
              setIsFocused({ ...isFocused, email: value })
            }
          />
          <PasswordField
            password={fields.password}
            setPassword={(value: string) =>
              setFields({ ...fields, password: value })
            }
            isFocused={isFocused}
            setIsFocused={(value: any) =>
              setIsFocused({ ...isFocused, password: value })
            }
          />
        </>
      )}
    </Form>
  );
};

export default SignUp;
