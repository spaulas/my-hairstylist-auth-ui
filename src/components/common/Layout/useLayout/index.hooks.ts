import { useSelector, useDispatch } from "react-redux";
import { RootReducerState } from "@type/global";
import layoutActions from "@actions/layout/layout.action";

const useLayout = () => {
  const dispatch = useDispatch();

  const { isBackgroundFocused } = useSelector(
    ({ Layout }: RootReducerState) => {
      return { isBackgroundFocused: Layout.isBackgroundFocused };
    }
  );

  const setBackgroundFocus = (focused: boolean) => {
    dispatch(layoutActions.setBackgroundFocus(focused));
  };

  return { isBackgroundFocused, setBackgroundFocus };
};

export default useLayout;
