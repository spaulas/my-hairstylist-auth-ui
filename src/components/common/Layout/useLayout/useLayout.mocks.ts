export const mockUseLayout = jest.fn();

jest.mock("./index.hooks.ts", () => mockUseLayout);
