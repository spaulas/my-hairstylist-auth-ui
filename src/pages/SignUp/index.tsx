import React from "react";
import Layout from "@components/common/Layout";
import SignUpForm from "@components/SignUp";

const SignUp = () => {
  return (
    <Layout>
      <SignUpForm />
    </Layout>
  );
};

export default SignUp;
