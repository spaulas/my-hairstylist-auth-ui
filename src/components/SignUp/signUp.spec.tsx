import "@testing-library/jest-dom/extend-expect";
import { mockUseState } from "../../__mocks__/mock-react";
import { mockUseAuth } from "@hooks/useAuth/useAuth.mocks";
import { mockUseLayout } from "@components/common/Layout/useLayout/useLayout.mocks";
import { fireEvent, render, screen } from "@testing-library/react";
import SignUp from ".";

describe("SignUp", () => {
  const mockSignUp = jest.fn();

  const fields = {
    password: "",
    email: "",
    name: "",
  };
  const mockSetFields = jest.fn();

  beforeEach(() => {
    mockUseAuth.mockReturnValue({ signUp: mockSignUp });
    mockUseState.mockReturnValue([fields, mockSetFields]);
    mockUseLayout.mockReturnValue({
      isBackgroundFocused: false,
      setBackgroundFocus: jest.fn(),
    });
  });

  it("renders correctly", () => {
    const { container } = render(<SignUp />);
    expect(container).toMatchSnapshot();
  });

  describe("renders correct fields", () => {
    beforeEach(() => {
      render(<SignUp />);
    });

    it("button", () => {
      const button = screen.getByText("Sign up");
      expect(button).toBeVisible();
    });

    it("forgotPassword", () => {
      const forgotPassword = screen.queryByText("Forgot your password?");
      expect(forgotPassword).toBeFalsy();
    });

    it("link message", () => {
      const linkMessage = screen.getByText("Already have an account?");
      expect(linkMessage).toBeVisible();
    });

    it("link title", () => {
      const linkTitle = screen.getByText("Sign in");
      expect(linkTitle).toBeVisible();
    });

    it("email field", () => {
      const emailTitle = screen.getByText("Email");
      expect(emailTitle).toBeVisible();
    });

    it("password field", () => {
      const passwordTitle = screen.getByText("Password");
      expect(passwordTitle).toBeVisible();
    });

    it("name field", () => {
        const nameTitle = screen.getByText("Name");
        expect(nameTitle).toBeVisible();
      });
  });

  it("link should have correct url", () => {
    render(<SignUp />);
    const linkTitle = screen.getByText("Sign in");
    expect(linkTitle).toHaveAttribute("href", "/auth/sign-in");
  });

  describe("should call setFields when", () => {
    it("email field changes", () => {
      render(<SignUp />);

      const emailInput = screen.getByTestId("email");
      const email = "email@email.com";
      fireEvent.change(emailInput, { target: { value: email } });

      expect(mockSetFields).toHaveBeenCalledWith({ ...fields, email });
    });

    it("password field changes", () => {
      render(<SignUp />);

      const passwordInput = screen.getByTestId("password");
      const password = "email@email.com";
      fireEvent.change(passwordInput, { target: { value: password } });

      expect(mockSetFields).toHaveBeenCalledWith({ ...fields, password });
    });


    it("name field changes", () => {
        render(<SignUp />);
  
        const nameInput = screen.getByTestId("name");
        const name = "Name";
        fireEvent.change(nameInput, { target: { value: name } });
  
        expect(mockSetFields).toHaveBeenCalledWith({ ...fields, name });
      });
  });

  it("calls signUp when button is clicked", () => {
    render(<SignUp />);

    const button = screen.getByText("Sign up");
    fireEvent.click(button);

    expect(mockSignUp).toHaveBeenCalledTimes(1);
  });
});
