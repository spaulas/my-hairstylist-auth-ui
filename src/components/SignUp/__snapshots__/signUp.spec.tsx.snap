// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp renders correctly 1`] = `
<div>
  <div
    class="form"
  >
    <h1
      class="form--title "
    >
      My Hairstylist
    </h1>
    <div
      class="form--item"
    >
      <p
        class="form--label form--label__focused"
      >
        Name
      </p>
      <input
        class="form--input form--input__focused"
        data-testid="name"
        id="name"
        name="name"
        type="name"
        value=""
      />
      <p
        class="error "
      />
    </div>
    <div
      class="form--item"
    >
      <p
        class="form--label form--label__focused"
      >
        Email
      </p>
      <input
        class="form--input form--input__focused"
        data-testid="email"
        id="email"
        name="email"
        type="email"
        value=""
      />
      <p
        class="error error__visible"
      >
        Please add a valid email
      </p>
    </div>
    <div
      class="form--item"
    >
      <p
        class="form--label form--label__focused"
      >
        Password
      </p>
      <input
        class="form--input form--input__focused"
        data-testid="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <p
        class="error error__visible"
      />
    </div>
    <button
      class="primary"
    >
      Sign up
    </button>
    <p
      class="form--links-sign-up"
    >
      Already have an account?
       Â 
      <a
        class="link "
        href="/auth/sign-in"
      >
        Sign in
      </a>
    </p>
  </div>
</div>
`;
